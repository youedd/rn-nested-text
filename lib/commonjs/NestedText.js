"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _reactNative = require("react-native");

var _parser = require("./parser");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

const NestedText = _ref => {
  let {
    children,
    textProps,
    ...props
  } = _ref;

  const parsed = _react.default.useMemo(() => {
    return _parser.Parser.parse(children);
  }, [children]);

  return /*#__PURE__*/_react.default.createElement(_reactNative.Text, props, renderNestedText(parsed, textProps));
};

const renderNestedText = function (nestedText) {
  let textProps = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  return nestedText.map((text, index) => {
    if (typeof text === 'string') {
      return text;
    }

    const props = textProps[text.tag];
    return /*#__PURE__*/_react.default.createElement(_reactNative.Text, _extends({
      key: index
    }, props), renderNestedText(text.children, textProps));
  });
};

var _default = NestedText;
exports.default = _default;
//# sourceMappingURL=NestedText.js.map