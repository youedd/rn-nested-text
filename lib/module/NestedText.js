function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

import React from 'react';
import { Text } from 'react-native';
import { Parser } from './parser';

const NestedText = _ref => {
  let {
    children,
    textProps,
    ...props
  } = _ref;
  const parsed = React.useMemo(() => {
    return Parser.parse(children);
  }, [children]);
  return /*#__PURE__*/React.createElement(Text, props, renderNestedText(parsed, textProps));
};

const renderNestedText = function (nestedText) {
  let textProps = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  return nestedText.map((text, index) => {
    if (typeof text === 'string') {
      return text;
    }

    const props = textProps[text.tag];
    return /*#__PURE__*/React.createElement(Text, _extends({
      key: index
    }, props), renderNestedText(text.children, textProps));
  });
};

export default NestedText;
//# sourceMappingURL=NestedText.js.map