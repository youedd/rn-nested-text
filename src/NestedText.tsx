import React from 'react'
import { Text, TextProps } from 'react-native'
import { Parser, TextGroup } from './parser'

type TextsProps = Record<string, TextProps> | undefined
export interface NestedTextProps extends TextProps {
  children: string
  textProps?: TextsProps
}

type NestedTextComponent = React.FC<NestedTextProps> & {
  defaultTextProps: Record<string, TextProps>
}

const DEFAULT_TEXT_PROPS: Record<string, TextProps> = {
  i: {
    style: {
      fontStyle: 'italic'
    }
  },
  u: {
    style: {
      textDecorationLine: 'underline'
    }
  },
  b: {
    style: {
      fontWeight: 'bold'
    }
  }
}

const NestedText: NestedTextComponent = ({ children, textProps, ...props }) => {
  const parsed = React.useMemo(() => {
    return Parser.parse(children)
  }, [children])

  return <Text {...props}>{renderNestedText(parsed, textProps)}</Text>
}

const renderNestedText = (
  nestedText: Array<string | TextGroup>,
  textProps: TextsProps = {}
): Array<string | JSX.Element> => {
  return nestedText.map((text, index) => {
    if (typeof text === 'string') {
      return text
    }

    const props = {
      ...NestedText.defaultTextProps[text.tag],
      ...textProps[text.tag]
    }

    return (
      <Text key={index} {...props}>
        {renderNestedText(text.children, textProps)}
      </Text>
    )
  })
}

NestedText.defaultTextProps = DEFAULT_TEXT_PROPS
export default NestedText
