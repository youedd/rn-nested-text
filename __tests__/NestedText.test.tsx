import React from 'react'
import { render } from '@testing-library/react-native'
import NestedText from '../src/NestedText'

it('should render simple text', () => {
  const { toJSON } = render(<NestedText>This is a Test</NestedText>)
  expect(toJSON()).toMatchSnapshot()
})

it('should render simple text with props', () => {
  const onPressMock = jest.fn()
  const { toJSON } = render(
    <NestedText
      testID='text'
      style={{ color: 'red' }}
      onPress={() => onPressMock()}
    >
      This is a Test
    </NestedText>
  )

  expect(toJSON()).toMatchSnapshot()
})

it('should render nested text ', () => {
  const { toJSON } = render(
    <NestedText>{'This <b>is a <a>Link</a></b>'}</NestedText>
  )

  expect(toJSON()).toMatchSnapshot()
})

it('should render nested text with props', () => {
  const onPressMock = jest.fn()
  const { toJSON } = render(
    <NestedText
      testID='text'
      style={{ color: 'red' }}
      textProps={{
        b: {
          testID: 'bold'
        },
        a: {
          testID: 'link',
          style: { color: 'blue', textDecorationLine: 'underline' },
          onPress: () => onPressMock()
        }
      }}
    >
      {'This <b>is a <a>Link</a></b>'}
    </NestedText>
  )

  expect(toJSON()).toMatchSnapshot()
})
