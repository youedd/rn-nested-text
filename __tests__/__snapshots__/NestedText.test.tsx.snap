// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render nested text  1`] = `
<Text>
  This 
  <Text>
    is a 
    <Text>
      Link
    </Text>
  </Text>
</Text>
`;

exports[`should render nested text with props 1`] = `
<Text
  style={
    Object {
      "color": "red",
    }
  }
  testID="text"
>
  This 
  <Text
    testID="bold"
  >
    is a 
    <Text
      onPress={[Function]}
      style={
        Object {
          "color": "blue",
          "textDecorationLine": "underline",
        }
      }
      testID="link"
    >
      Link
    </Text>
  </Text>
</Text>
`;

exports[`should render simple text 1`] = `
<Text>
  This is a Test
</Text>
`;

exports[`should render simple text with props 1`] = `
<Text
  onPress={[Function]}
  style={
    Object {
      "color": "red",
    }
  }
  testID="text"
>
  This is a Test
</Text>
`;
